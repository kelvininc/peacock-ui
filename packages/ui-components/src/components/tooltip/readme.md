# *<kv-tooltip>*


<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-tooltip text="Tooltip">
	<kv-action-button [type]="EActionButtonType.Primary">Hover me!</kv-action-button>
</kv-tooltip>

<!-- With Fixed Position -->
<kv-tooltip text="Tooltip" [position]="ETooltipPosition.Left">
	<kv-action-button [type]="EActionButtonType.Primary">Hover me!</kv-action-button>
</kv-tooltip>

<!-- With Allowed Position -->
<kv-tooltip text="Tooltip" [allowedPositions]="[ETooltipPosition.Top, ETooltipPosition.Bottom]">
	<kv-action-button [type]="EActionButtonType.Primary">Hover me!</kv-action-button>
</kv-tooltip>
```


### React

```tsx
import React from 'react';

import { KvTooltip, KvActionButton, ETooltipPosition, EActionButtonType } from '@kelvininc/react-ui-components';

export const TagLetterExample: React.FC = () => (
  <>
    {/*-- Default --*/}
	<KvTooltip text="Tooltip">
		<KvActionButton type={EActionButtonType.Primary}>Hover me!</KvActionButton>
	</KvTooltip>

	{/*-- With Fixed Position --*/}
	<KvTooltip text="Tooltip" position={ETooltipPosition.Left}>
		<KvActionButton type={EActionButtonType.Primary}>Hover me!</KvActionButton>
	</KvTooltip>

	{/*-- With Allowed Positions --*/}
	<KvTooltip text="Tooltip" allowedPositions={[ETooltipPosition.Top, ETooltipPosition.Bottom]}>
		<KvActionButton type={EActionButtonType.Primary}>Hover me!</KvActionButton>
	</KvTooltip>
  </>
);
```



## Properties

| Property         | Attribute  | Description                                     | Type                                                                                                                                                                                                                                                                                                                                             | Default                   |
| ---------------- | ---------- | ----------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | ------------------------- |
| `contentElement` | --         | (optional) Content element reference            | `HTMLElement`                                                                                                                                                                                                                                                                                                                                    | `null`                    |
| `disabled`       | `disabled` | (optional) Disables tooltip                     | `boolean`                                                                                                                                                                                                                                                                                                                                        | `false`                   |
| `options`        | --         | (optional) Object with tooltip position options | `{ platform?: Platform; placement?: Placement; strategy?: Strategy; middleware?: Middleware[]; }`                                                                                                                                                                                                                                                | `DEFAULT_POSITION_CONFIG` |
| `position`       | `position` | (optional) Position of tooltip                  | `ETooltipPosition.Bottom \| ETooltipPosition.BottomEnd \| ETooltipPosition.BottomStart \| ETooltipPosition.Left \| ETooltipPosition.LeftEnd \| ETooltipPosition.LeftStart \| ETooltipPosition.Right \| ETooltipPosition.RightEnd \| ETooltipPosition.RightStart \| ETooltipPosition.Top \| ETooltipPosition.TopEnd \| ETooltipPosition.TopStart` | `undefined`               |
| `text`           | `text`     | (optional) Text of tooltip                      | `string`                                                                                                                                                                                                                                                                                                                                         | `undefined`               |


## Shadow Parts

| Part          | Description            |
| ------------- | ---------------------- |
| `"container"` | The tooltip container. |
| `"content"`   | The tooltip content.   |


## CSS Custom Properties

| Name                  | Description                                  |
| --------------------- | -------------------------------------------- |
| `--container-z-index` | The z-index value for the tooltip container. |


## Dependencies

### Used by

 - [kv-info-label](../info-label)
 - [kv-text-field](../text-field)

### Graph
```mermaid
graph TD;
  kv-info-label --> kv-tooltip
  kv-text-field --> kv-tooltip
  style kv-tooltip fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


