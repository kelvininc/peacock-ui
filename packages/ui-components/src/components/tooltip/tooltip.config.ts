import { ComputePositionConfig, offset, shift } from '@floating-ui/dom';

export const DEFAULT_POSITION_CONFIG: Partial<ComputePositionConfig> = {
	strategy: 'fixed',
	middleware: [offset(5), shift({ padding: 5 })]
};
