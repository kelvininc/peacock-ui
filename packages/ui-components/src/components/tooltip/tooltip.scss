@import '../../assets/styles/globals';

:host {
	/**
	  * @prop --container-z-index: The z-index value for the tooltip container.
	 */

	--container-z-index: 10;
}

.tooltip-content {
	width: fit-content;
	height: fit-content;
}

.tooltip-container {
	@include kv-font-span-semibold();
	height: fit-content;
	width: fit-content;
	max-width: 240px;
	word-wrap: break-word;
	color: kv-color('neutral-0');
	background-color: kv-color('neutral-6');
	box-shadow: 0 8px 16px -4px kv-color('neutral-6', 'base', 0.25), 0 0 1px 0 kv-color('neutral-3', 'base', 0.31);
	pointer-events: none;
	user-select: none;
	border-radius: 2px;
	padding: $spacing $spacing-2x;
	position: fixed;
	z-index: var(--container-z-index);
	display: none;
}
