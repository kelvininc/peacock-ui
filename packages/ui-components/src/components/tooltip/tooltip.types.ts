import { ComputePositionConfig } from '@floating-ui/dom';
import { ETooltipPosition } from '../../types';

export interface ITooltip {
	/** (optional) Text of tooltip */
	text: string;
	/** (optional) Position of tooltip */
	position?: ETooltipPosition;
	/** (optional) Object with tooltip position options */
	options?: Partial<ComputePositionConfig>;
	/** (optional) Disables tooltip */
	disabled?: boolean;
	/** (optional) Content element reference */
	contentElement?: HTMLElement;
}
