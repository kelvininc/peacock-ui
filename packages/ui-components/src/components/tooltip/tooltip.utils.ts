import { getCollapsedElement } from '../utils';

export const isElementCollpased = (element: HTMLElement): boolean => {
	const collapsedElement = getCollapsedElement(Array.from(element.childNodes));
	if (collapsedElement !== undefined) {
		const range = new Range();
		range.selectNodeContents(collapsedElement);

		const { width: rectWidth } = range.getBoundingClientRect();
		const { clientWidth: collapsedWidth } = collapsedElement;

		// once rectWidth comes as a real value and collapsedWidth as a rounded value, this calculation was incorrect.
		if (collapsedWidth >= Math.round(rectWidth)) return false;
	}

	return true;
};
