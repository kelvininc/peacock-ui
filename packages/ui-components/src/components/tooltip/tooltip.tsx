import { Component, Element, Host, Prop, State, h } from '@stencil/core';
import { ComputePositionConfig, computePosition } from '@floating-ui/dom';
import { isEmpty, merge } from 'lodash-es';

import { DEFAULT_DELAY_CONFIG, DEFAULT_POSITION_CONFIG } from './tooltip.config';
import { ETooltipPosition } from '../../types';
import { ITooltip } from './tooltip.types';
import { isElementCollpased } from './tooltip.utils';

/**
 * @part container - The tooltip container.
 * @part content - The tooltip content.
 */
@Component({
	tag: 'kv-tooltip',
	styleUrl: 'tooltip.scss',
	shadow: true
})
export class KvTooltip implements ITooltip {
	/** (optional) Delay to show tooltip in milliseconds. */
	@Prop({ reflect: true }) delay?: number = DEFAULT_DELAY_CONFIG;
	/** @inheritdoc */
	@Prop({ reflect: true }) text: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) position?: ETooltipPosition;
	/** @inheritdoc */
	@Prop({ reflect: false }) options?: Partial<ComputePositionConfig> = DEFAULT_POSITION_CONFIG;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled?: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: false }) contentElement?: HTMLElement = null;
	/** @inheritdoc */
	@Prop({ reflect: false }) truncate?: boolean = false;

	@State() timeoutDelayId?: number;

	/** The Host's element reference */
	@Element() el: HTMLKvTooltipElement;

	private getOptions = (): Partial<ComputePositionConfig> => {
		return merge({}, { placement: this.position }, this.options);
	};

	private getTooltipElement = (): HTMLElement | null => {
		return this.el.shadowRoot.querySelector('#tooltip') as HTMLElement | null;
	};

	private getContentElement = (): HTMLElement | null => {
		return this.contentElement ?? (this.el.shadowRoot.querySelector('#content') as HTMLElement | null);
	};

	private showTooltip = () => {
		if (!this.disabled) {
			const tooltip = this.getTooltipElement();
			tooltip.style.display = 'inline-block';
			this.update();
		}
	};

	private showTooltipHandler = () => {
		if (this.truncate && !isElementCollpased(this.el)) return;

		if (this.delay) {
			this.timeoutDelayId = window.setTimeout(() => {
				this.showTooltip();
			}, this.delay);
		} else {
			this.showTooltip();
		}
	};

	private hideTooltip = () => {
		const tooltip = this.getTooltipElement();
		tooltip.style.display = '';
	};

	private hideTooltipHandler = () => {
		this.hideTooltip();
		if (this.timeoutDelayId) {
			clearTimeout(this.timeoutDelayId);
			this.timeoutDelayId = undefined;
		}
	};

	private update = () => {
		const tooltip = this.getTooltipElement();
		const child = this.getContentElement();

		computePosition(child, tooltip, this.getOptions()).then(({ x, y }) => {
			tooltip.style.left = `${x}px`;
			tooltip.style.top = `${y}px`;
		});
	};

	private listenToEvents = (child: HTMLElement) => {
		child.addEventListener('mouseenter', this.showTooltipHandler);
		child.addEventListener('mouseleave', this.hideTooltipHandler);
		child.addEventListener('focus', this.showTooltip);
		child.addEventListener('blur', this.hideTooltip);
	};

	private unlistenToEvents = (child: HTMLElement) => {
		child.removeEventListener('mouseenter', this.showTooltipHandler);
		child.removeEventListener('mouseleave', this.hideTooltipHandler);
		child.removeEventListener('focus', this.showTooltip);
		child.removeEventListener('blur', this.hideTooltip);
	};

	componentDidRender() {
		const child = this.getContentElement();
		const tooltip = this.getTooltipElement();

		if (tooltip === null) {
			return;
		}
		this.unlistenToEvents(child);
		this.listenToEvents(child);
	}

	disconnectedCallback() {
		const child = this.getContentElement();
		if (child) {
			this.unlistenToEvents(child);
		}
	}

	render() {
		return (
			<Host>
				<div id="content" aria-describedby="tooltip" part="content">
					<slot></slot>
				</div>
				{!isEmpty(this.text) && (
					<div id="tooltip" class="tooltip-container" role="tooltip" part="container">
						{this.text}
					</div>
				)}
			</Host>
		);
	}
}
