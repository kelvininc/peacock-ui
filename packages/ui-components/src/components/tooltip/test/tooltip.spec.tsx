import { SpecPage } from '@stencil/core/internal';
import { KvTooltip } from '../tooltip';
import { newSpecPage } from '@stencil/core/testing';

describe('Tooltip (unit tests)', () => {
	let page: SpecPage;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTooltip],
				html: `<kv-tooltip text="Tooltip"></kv-tooltip>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when has a position', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvTooltip],
				html: `<kv-tooltip text="Tooltip" position="left"></kv-tooltip>`
			});
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});
});
