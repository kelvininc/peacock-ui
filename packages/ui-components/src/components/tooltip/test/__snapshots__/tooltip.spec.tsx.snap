// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip (unit tests) when has a position should match the snapshot 1`] = `
<kv-tooltip position="left" text="Tooltip">
  <mock:shadow-root>
    <div aria-describedby="tooltip" class="tooltip-content" id="content" part="content">
      <slot></slot>
    </div>
    <div class="tooltip-container" id="tooltip" part="container" role="tooltip">
      Tooltip
    </div>
  </mock:shadow-root>
</kv-tooltip>
`;

exports[`Tooltip (unit tests) when uses default props should match the snapshot 1`] = `
<kv-tooltip text="Tooltip">
  <mock:shadow-root>
    <div aria-describedby="tooltip" class="tooltip-content" id="content" part="content">
      <slot></slot>
    </div>
    <div class="tooltip-container" id="tooltip" part="container" role="tooltip">
      Tooltip
    </div>
  </mock:shadow-root>
</kv-tooltip>
`;
